frontend/server/src/main/java/org/pytorch/serve/ensemble/DagExecutor.java

```diff
@@ -0,0 +1,161 @@
+                    leafOutputs.add(output);
+                } else {
+                    for (String newNodeName : childNodes) {
+
+                        if (topoSortedList == null) {
+                            byte[] response = (byte[]) output.getData();
+
+                            RequestInput newInput = this.inputRequestMap.get(newNodeName);
+                            if (newInput == null) {
+                                List<InputParameter> params = new ArrayList<>();
+                                newInput = new RequestInput(UUID.randomUUID().toString());
+                                if (inDegreeMap.get(newNodeName) == 1) {
+                                    params.add(new InputParameter("body", response));
+                                } else {
+                                    params.add(new InputParameter(nodeName, response));
+                                }
+                                newInput.setParameters(params);
+                                newInput.setHeaders(input.getHeaders());
+                            } else {
+                                newInput.addParameter(new InputParameter(nodeName, response));
+                            }
+                            this.inputRequestMap.put(newNodeName, newInput);
+                        }
+
+                        inDegreeMap.replace(newNodeName, inDegreeMap.get(newNodeName) - 1);
+                        if (inDegreeMap.get(newNodeName) == 0) {
+                            zeroInDegree.add(newNodeName);
+                        }
+                    }
+                }
+            }
+        }
+
+        return leafOutputs;
+    }
+
+    private NodeOutput invokeModel(String nodeName, WorkflowModel workflowModel, RequestInput input)
+            throws ModelNotFoundException, ModelVersionNotFoundException, ExecutionException,
+                    InterruptedException {
+        try {
+            CompletableFuture<byte[]> respFuture = new CompletableFuture<>();
+
+            RestJob job = ApiUtils.addRESTInferenceJob(null, workflowModel.getName(), null, input);
+            job.setResponsePromise(respFuture);
+            byte[] resp = respFuture.get();
+            return new NodeOutput(nodeName, resp);
+        } catch (InterruptedException | ExecutionException e) {
+            logger.error("Failed to execute workflow Node.");
+            logger.error(nodeName + " : " + e.getMessage());
+            String[] error = e.getMessage().split(":");
+            throw new InternalServerException(nodeName + " - " + error[error.length - 1]); // NOPMD
+        } catch (ModelNotFoundException e) {
+            logger.error("Model not found.");
+            logger.error(e.getMessage());
+            throw e;
+        } catch (ModelVersionNotFoundException e) {
+            logger.error("Model version not found.");
+            logger.error(e.getMessage());
+            throw e;
+        }
+    }
+}
```

Comment:
ListenableFuture may be a more appropriate 

https://guava.dev/releases/21.0/api/docs/com/google/common/util/concurrent/ListenableFuture.html

https://stackoverflow.com/questions/19138212/how-to-implement-a-dag-like-scheduler-in-java has a good example.